### DISCUSSION TITLE
Passing additional custom data to LLMRails generate call that can be passed to actions

### DISCUSSION THREAD

**Discussion by avirgvd** [NVIDIA-NeMo/Guardrails#338] [Q&A]
I am looking for ways to send custom context information along with LLMRails generate() call. Need this additional context for my actions. How can I achieve this?

Below is my Python code passing my custom parameter "key=123". I expect the actions to access this parameter value. 

```python
        self.rails = LLMRails(guardrails_config)
        response = self.rails.generate(key=123, messages=[{
            "role": "user",
            "content": message
        }])
```

```colang
# INPUT RAILS
define subflow mask sensitive data on input
  """Mask any sensitive data found in the user input."""
  $user_message = execute mask_sensitive_data(source="input", text=$user_message, callid=$key)

# OUTPUT RAILS
define subflow unmask sensitive data on output
  """Unmask any masked data found in the bot output."""
  $bot_message = execute unmask_sensitive_data(source="output", text=$bot_message, callid=$key)

```

**Comment by drazvan** [comment]
Yes, you can. Use a message with `role` set to `context`. For example:
```json
[
  {
    "role": "context",
    "content": {
      "key": "123"
    }
  },
  {
    "role": "user",
    "content": "Hello!"
  }
]
```

**Reply by avirgvd** [reply]
Thank you very much for the answer. 
Adding context the way you suggested is working. Now I need to pass this context to the actions. How can I do this?

**Reply by avirgvd** [reply]
The contextvars approach did not work. Not sure if I coded right. It appeared that the contextvars is out of scope for the custom rails actions.py that I have implemented for LLM input and output.

**Reply by zhangl1201** [reply]
I have tried to put context as optional params in action, and execute action in flows.co. It looks like "execute" will automatically inject the context param and config param. And then you could use context.get("user_message") to get variables in context.

actions.py
```
@action(is_system_action=True)
async def encrypt_sensitive_data_input(
    config: RailsConfig,
    context: Optional[dict] = None
):
    """Checks whether the provided text contains any sensitive data and encrypt the sensitive data.

    Args
        config: The rails configuration object.
        context: context values.

    Returns
        The altered text, if applicable.
    """
    text = context.get("user_message")
    source = "input"
    # Based on the source of the data, we use the right options
    sdd_config = config.rails.config.sensitive_data_detection
    options: SensitiveDataDetectionOptions = getattr(sdd_config, source)
```

flows.co
```
define flow encrypt sensitive data on input
  """Mask any sensitive data found in the user input."""
  $user_message = execute encrypt_sensitive_data_input
```

**Reply by drazvan** [reply]
Yes, @zhangl1201 is right. Adding a parameter named `context` for an action will give it access to the context. So, if you pass the context as part of the input messages, you can also access it directly from actions as well.
